core/lens/lens_database.ml

```diff
@@ -0,0 +1,286 @@
+  Format.fprintf f "%a" (fmt_table ~db |> fmt_comma_seperated) tables
+
+let fmt_cols ~db f cols =
+  Format.fprintf f "%a" (fmt_col ~db |> fmt_comma_seperated) cols
+
+let rec fmt_phrase ~db ~map f expr =
+  let pp_sep f () = Format.fprintf f ", " in
+  let fmt = fmt_phrase ~db ~map in
+  match expr with
+  | Constant c -> Format.fprintf f "%a" Constant.fmt c
+  | Var v -> Format.fprintf f "%s" (map v)
+  | InfixAppl (op, a1, a2) -> Format.fprintf f "%a %a %a" fmt a1 Binary.fmt op fmt a2
+  | TupleLit l -> Format.fprintf f "(%a)" (Format.pp_print_list ~pp_sep fmt) l
+  | UnaryAppl (op, a) ->
+    let op = match op with
+      | Unary.Not -> "NOT"
+      | _ -> Unary.to_string op in
+    Format.fprintf f "%s (%a)" op fmt a
+  | In (names, vals) ->
+    let fmt_name f v = Format.fprintf f "%s" (map v) in
+    let fmt_val f v = Format.fprintf f "(%a)" (Format.pp_print_list ~pp_sep Constant.fmt) v in
+    (match vals with
+     | [] -> Format.fprintf f "FALSE"
+     | _ ->
+       Format.fprintf f "(%a) IN (%a)"
+         (Format.pp_print_list ~pp_sep fmt_name) names
+         (Format.pp_print_list ~pp_sep fmt_val) vals)
+  | Case (inp, cases, otherwise) ->
+    if cases = [] then
+      Format.fprintf f "%a" fmt otherwise
+    else
+      let f =
+        match inp with
+        | None -> Format.fprintf f "CASE %a ELSE %a END"
+        | Some inp -> Format.fprintf f "CASE (%a) %a ELSE %a END" fmt inp in
+      let pp_sep f () = Format.fprintf f " " in
+      let fmt_case f (k,v) = Format.fprintf f "WHEN %a THEN %a" fmt k fmt v in
+      f (Format.pp_print_list ~pp_sep fmt_case) cases fmt otherwise
+
+let fmt_phrase_dummy f expr =
+  let db = (new dummy_database) in
+  let map = fun a -> db#quote_field a in
+  fmt_phrase ~db ~map f expr
+
+let to_string_dummy expr =
+  Format.asprintf "%a" fmt_phrase_dummy expr
+
+module Select = struct
+  type db = t
+
+  type t = {
+    tables : (string * string) list;
+    cols : lens_col list;
+    predicate: lens_phrase option;
+    db : db;
+  }
+
+  let of_sort t ~sort =
+    let predicate = Lens_sort.predicate sort in
+    let cols = Lens_sort.cols sort in
+    let tables =
+      List.map Lens_column.table cols
+      |> List.sort_uniq String.compare
+      |> List.map (fun c -> (c,c)) in
+    {
+      predicate;
+      cols;
+      tables;
+      db = t;
+    }
+
+  let fmt f v =
+    let db = v.db in
+    let map = fun a ->
+      let col = List.find (fun c -> c.alias = a) v.cols in
+      Format.sprintf "%s.%s" (db#quote_field col.table) (db#quote_field col.name) in
+    let cols = v.cols |> Lens_column.List.present in
+    match v.predicate with
+    | None -> Format.fprintf f "SELECT %a FROM %a" (fmt_cols ~db) cols (fmt_tables ~db) v.tables
+    | Some pred ->
+      Format.fprintf f "SELECT %a FROM %a WHERE %a"
+        (fmt_cols ~db) cols
+        (fmt_tables ~db) v.tables
+        (fmt_phrase ~db ~map) pred
+
+  (* execution, as taken from database.ml (this cannot be reused though, as
```

Comment:
What is the dependency relation between this module and `database.ml`?

Reply:
This seems to be an issue that only existed with the old code. As such I have removed these copied functions and now refer to the `Database` module.